lib/Sema/TypeCheckGeneric.cpp

```diff
@@ -81,6 +81,24 @@ TypeChecker::gatherGenericParamBindingsText(
 // Generic functions
 //
 
+// Get a serialized opaque type corresponding to a naming decl.
+OpaqueTypeDecl *
+OpaqueResultTypeFromSerializedASTRequest::evaluate(Evaluator &evaluator,
+                                                   ValueDecl *vd) const {
+  for (auto *file: vd->getModuleContext()->getFiles()) {
```

Comment:
It would be nice to avoid the O(n) iteration, and full deserialization of all decls that have an opaque result here.

Instead, you can look up the type via the mangled name, by calling `ModuleFile::lookupOpaqueResultType()`. You would pass in the result of `ASTMangler::mangleOpaqueTypeDecl()` (however, you can add a new overload of the latter taking a ValueDecl and not the OpaqueTypeDecl. The only thing it uses the OpaqueTypeDecl for is getting the ValueDecl out via `getNamingDecl()').

If you do that, you probably don't even need to define a new request kind here.

Reply:
hmm, I'm confused. At this point, we only have access to the naming decl `vd`.  How can we get the mangled name of it to look up by calling `ModuleFile::lookupOpaqueResultType()`?

Reply:
Ah, that makes sense. Updated in the latest commit.